# Учебный проект 2_Анализ оттока клиентов банка

## Содержание

* [Описание проекта](#Описание)
* [Импорт библиотек Python и загрузка данных](#Импорт)
* [Предобработка данных](#Предобработка)

## Описание проекта <a class = 'anchor' id = 'Описание'></a>

## Импорт библиотек Python и загрузка данных <a class = 'anchor' id = 'Импорт'></a>

Данный блок характеризуется следующими последовательными действиями:

1. Импорт библиотек Python:
    * для манипулирования данными;
    * для визуализации данных;
2. Загрузка данных в рабочую среду Jupyter Notebook. Инициализация переменной **data** для хранения загруженного датасета;
3. Вывод на экран параметров датасета:
    * Перечень столбцов набора данных;
    * Количество строк в наборе данных;
    * Количество столбцов в наборе данных.

In [15]:
# импорт библиотек Python
import pandas as pd #для анализа табличных данных
import numpy as np #для работы с объектами линейной алгебры

import matplotlib.pyplot as plt # для визуализации данных

In [3]:
# инициализация переменной 'data' и загрузка набора данных
data = pd.read_csv('~\Desktop\YandexPraktikum_projects\datasets\data.csv')

# вывод на экран перечня столбцов датасета
for column in data.columns.to_list():
    print('*', column)

print()
print('Количество строк в наборе данных', data.shape[0])
print('Количество столбцов в наборе данных', data.shape[1])

* children
* days_employed
* dob_years
* education
* education_id
* family_status
* family_status_id
* gender
* income_type
* debt
* total_income
* purpose

Количество строк в наборе данных 21525
Количество столбцов в наборе данных 12


<>:2: SyntaxWarning: invalid escape sequence '\D'
<>:2: SyntaxWarning: invalid escape sequence '\D'
C:\Users\k.storozhuk\AppData\Local\Temp\ipykernel_24204\1267969003.py:2: SyntaxWarning: invalid escape sequence '\D'
  data = pd.read_csv('~\Desktop\YandexPraktikum_projects\datasets\data.csv')


**Вывод:**

1. Импортированы библиотеки Python:
    * для работы с данными табличного вида;
    * для визуализации данных.
2. Инициализирована переменная **data** и загружен рабочий датасет в рабочую среду;
3. Выведены на экран следующие параметры датасета:
    * перечень названий столбцов с данными;
    * количество строк в наборе данных;
    * количество столбцов в наборе данных.

## Предобработка данных <a class = 'anchor' id = 'Предобработка'></a>

Данный блок характеризуется следующими последовательными действиями:

1. Проверка набора данных на пропущенные значения. Принятие решения об исключении пропущенных значений в датасете;
2. Проверка набора даных на наличие аномальных значений:
    * Проверка на аномально большие значения;
    * Проверка на аномально малые значения;
3. Изучение датасета на наличие дубликатов (явных и неявных). Выбор метода по исключению дубликатов из набора данных;

In [33]:
# проверка типов данных в датасете
data.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

In [4]:
# вывод на экран количества пустых значений в наборе данных
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [10]:
# вывод на экран доли пропущенных значений от общего количества данных в наборе
print(f'Доля пропущенных значений в наборе данных: {(data['days_employed'].isna().sum() / data.shape[0]):.1%}')

Доля пропущенных значений в наборе данных: 10.1%


In [ ]:
# заполнение пустых ячеек столбца 'total_income' медианным значением
data['total_income'] = data['total_income'].fillna(data['total_income'].median())

In [25]:
# вывод на экран описательной статистики по набору данных
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод по промежуточному этапу:**

Обнаружены аномальные значения в следующих показателях:

    * 'children': минимальное значение "-1", максимальное значение "20";
    * 'days_employed': многочисленные отрицательные значения.

In [32]:
data[data['children'] < 0]['children'].sum()

-47

**Вывод:**

1. Набор проверен на пропущенные значения. Пустые ячейки обнаружены в следующих столбцах:
    * 'days_employed' - 2 174 значения. Что составляет **10.1%** от общего количества строк;
    * 'total_income' - 2 174 значения. Что составляет **10.1%** от общего количества строк;

Среди причин появления могут быть технические ошибки при формировании набора данных, а также человеческий фактор - ошибки при внесении данных в базу, а также сам факт не указания информации клиентом о себе.

Используя вводную из описания задания, пустые значения в столбце 'total_income' были заполнены медианным значением по данному столбцу.

Заполнение пустых ячеек медианным значением приоритетнее среднего по причине того, что медиана, как статистическая метрика, меньше подвержена влиянию аномальных значений.
